### ISSUE TITLE
OpenBLAS check thwarts Cray compiler wrappers

### ISSUE THREAD

**Issue by bcfriesen** [project-asgard/asgard#156]
Compiler wrappers on Cray XC systems (`cc`, `CC`, `ftn`) include appropriate include and link flags for BLAS/LAPACK implementations provided by Cray LibSci. These libraries sometimes have obscure names, e.g., `libsci_intel_mpi.so`. However, the OpenBLAS check in `contrib/contrib.cmake` makes it difficult to convince CMake that the BLAS requirement in ASGard is satisfied without providing explicit links to a linear algebra library.

Would it be possible to add a 'last-resort' mechanism to the BLAS check during configuration which simply attempts to compile a BLAS code with the given compiler and compiler flags? If that fails, then CMake could proceed to download OpenBLAS. Currently the way I work around this on the Cori system at NERSC is by removing the entire BLAS check in `contrib/contrib.cmake`.

**Comment by grahamlopez** [comment]
Did you try `-DASGARD_BLAS_PATH` or `-DASGARD_LAPACK_PATH`? I'm not sure that it would be sufficient, but we should understand the limitation.
I think a better 'last-resort' (if necessary) might be to allow the above options to take the full path to a file, including the file name. Then the responsibility moves to the user to ensure that whatever this file is provides the necessary symbols to compile/link.
Does this sound reasonable?

**Comment by bcfriesen** [comment]
Hi @grahamlopez,

> Did you try -DASGARD_BLAS_PATH or -DASGARD_LAPACK_PATH? 

Trying either/both of those CMake flags did not work.

> I think a better 'last-resort' (if necessary) might be to allow the above options to take the full path to a file, including the file name. Then the responsibility moves to the user to ensure that whatever this file is provides the necessary symbols to compile/link.
Does this sound reasonable?

I'm not sure this is the best approach. The problem is that LibSci provides different implementations based on whether the code is serial, threaded, or MPI-parallel. The compiler wrapper figures out which one to link to your program based on the flags that you provide. E.g., with the Intel compiler:

```
friesen@cori12:dgemm> cc dgemm_libsci.c
friesen@cori12:dgemm> ldd a.out|grep libsci
        libsci_intel.so.5 => /opt/cray/pe/lib64/libsci_intel.so.5 (0x00002aaaab52e000)
friesen@cori12:dgemm> cc -qopenmp dgemm_libsci.c
friesen@cori12:dgemm> ldd a.out|grep libsci
        libsci_intel_mp.so.5 => /opt/cray/pe/lib64/libsci_intel_mp.so.5 (0x00002aaaab52e000)
friesen@cori12:dgemm>
```

So then compiling ASGarD would require the user not only to set the correct compiler flags, but also set the full path to the appropriate LibSci library. This sounds prone to error.

**Comment by grahamlopez** [comment]
Ah yes, I had slowly forgotten the mess that is cray-pe. I agree with your comments.

**Comment by stonecoldhughes** [comment]
If I understand this issue correctly...
cmake: do you have blas? If yes, link to it. If no, download, build, and link.
That approach fails in the situation that a BLAS implementation exists and is already linked in, but cmake thinks it needs to explicitly link and the file has a weird name that it can't find.
The current command line CMake options are "Build and link" or "link but do not build". If we added a third option of "do not build and do not link because it's already taken care of", could that fix the problem?

**Comment by grahamlopez** [comment]
> third option of "do not build and do not link because it's already taken care of", could that fix the problem?

Yes, this might be the nuclear option. I was hoping that before we bring that one out, we could detect usage of the cray compilers and allow cmake to bypass our blas handling for the cray-pe case alone.

**Comment by bcfriesen** [comment]
I learned that CMake 3.16 has solved the FindBLAS problem on Cray's, although it has not yet solved the corresponding FindLAPACK problem. But even without 3.16 I think there is a workaround for this. On a plane now, but I will test and report back later today.

**Comment by bmcdanie** [comment]
@bcfriesen did you have a chance to look at this?

**Comment by bcfriesen** [comment]
Hi, happy new year. I tried a stupid hack which did not work. However, it seems the `FindBLAS` and `FindLAPACK` built-in functions were both fixed in CMake 3.16 such that they handle Cray LibSci correctly. E.g., in CMake 3.10.2:

```console
friesen@cori09:build> /usr/bin/cmake ..
-- The C compiler identification is Clang 9.0.0
-- The CXX compiler identification is Clang 9.0.0
-- Check for working C compiler: /opt/cray/pe/craype/2.6.2/bin/cc
-- Check for working C compiler: /opt/cray/pe/craype/2.6.2/bin/cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: /opt/cray/pe/craype/2.6.2/bin/CC
-- Check for working CXX compiler: /opt/cray/pe/craype/2.6.2/bin/CC -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- The Fortran compiler identification is Cray 9.1
-- Cray Programming Environment 2.6.2 Fortran
-- Check for working Fortran compiler: /opt/cray/pe/craype/2.6.2/bin/ftn
-- Check for working Fortran compiler: /opt/cray/pe/craype/2.6.2/bin/ftn  -- works
-- Detecting Fortran compiler ABI info
-- Detecting Fortran compiler ABI info - done
-- Checking whether /opt/cray/pe/craype/2.6.2/bin/ftn supports Fortran 90
-- Checking whether /opt/cray/pe/craype/2.6.2/bin/ftn supports Fortran 90 -- yes
-- Looking for pthread.h
-- Looking for pthread.h - found
-- Looking for pthread_create
-- Looking for pthread_create - found
-- Found Threads: TRUE
-- A library with BLAS API not found. Please specify library location.
-- A library with BLAS API not found. Please specify library location.
-- LAPACK requires BLAS
-- A library with LAPACK API not found. Please specify library location.
-- Configuring done
-- Generating done
-- Build files have been written to: /global/homes/f/friesen/tests/CMake/libsci/build
friesen@cori09:build>
```

But in the 3.16 dev build fresh from Kitware's GitLab repo:

```console
friesen@cori09:build> /global/cscratch1/sd/friesen/cmake.git/build/test_install/bin/cmake --version
cmake version 3.16.20200102-g98272cd

CMake suite maintained and supported by Kitware (kitware.com/cmake).
friesen@cori09:build> /global/cscratch1/sd/friesen/cmake.git/build/test_install/bin/cmake ..
-- The C compiler identification is Clang 9.0.0
-- The CXX compiler identification is Clang 9.0.0
-- Cray Programming Environment 2.6.2 C
-- Check for working C compiler: /opt/cray/pe/craype/2.6.2/bin/cc
-- Check for working C compiler: /opt/cray/pe/craype/2.6.2/bin/cc - works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Cray Programming Environment 2.6.2 CXX
-- Check for working CXX compiler: /opt/cray/pe/craype/2.6.2/bin/CC
-- Check for working CXX compiler: /opt/cray/pe/craype/2.6.2/bin/CC - works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- The Fortran compiler identification is Cray 9.1
-- Cray Programming Environment 2.6.2 Fortran
-- Check for working Fortran compiler: /opt/cray/pe/craype/2.6.2/bin/ftn
-- Check for working Fortran compiler: /opt/cray/pe/craype/2.6.2/bin/ftn - works
-- Detecting Fortran compiler ABI info
-- Detecting Fortran compiler ABI info - done
-- Checking whether /opt/cray/pe/craype/2.6.2/bin/ftn supports Fortran 90
-- Checking whether /opt/cray/pe/craype/2.6.2/bin/ftn supports Fortran 90 - yes
-- Looking for Fortran sgemm
-- Looking for Fortran sgemm - found
-- Found BLAS: BLAS_LIBRARIES-PLACEHOLDER-FOR-EMPTY-LIBRARIES
-- Looking for pthread.h
-- Looking for pthread.h - found
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Success
-- Found Threads: TRUE
-- Looking for Fortran cheev
-- Looking for Fortran cheev - found
-- A library with LAPACK API found.
-- Configuring done
-- Generating done
-- Build files have been written to: /global/homes/f/friesen/tests/CMake/libsci/build
friesen@cori09:build>
```
But it seems ASGarD does not use simply `find_package(LAPACK)` and `find_package(BLAS)`, and if it is desired not to use those, then we will have to implement a manual fix. What do you think?

**Comment by bcfriesen** [comment]
Never mind, I see it does use those `find_package()` commands. But the build gives up on LAPACK anyway and still tries to build OpenBLAS.

**Comment by grahamlopez** [comment]
I'm not sure what you mean by "the build gives up on LAPACK anyway and still tries to build OpenBLAS."? Looking back at it, the intended behavior is that if `find_package (LAPACK QUIET)` succeeds, that is what it uses.

**Comment by bcfriesen** [comment]
Whoops, forgot to include output in that last comment:

```console
friesen@cori09:build_cce9> /global/cscratch1/sd/friesen/cmake.git/build/test_install/bin/cmake --version
cmake version 3.16.20200102-g98272cd

CMake suite maintained and supported by Kitware (kitware.com/cmake).
friesen@cori09:build_cce9> /global/cscratch1/sd/friesen/cmake.git/build/test_install/bin/cmake -D CMAKE_CXX_COMPILER=CC ..
-- The CXX compiler identification is Clang 9.0.0
-- Cray Programming Environment 2.6.2 CXX
-- Check for working CXX compiler: /opt/cray/pe/craype/2.6.2/bin/CC
-- Check for working CXX compiler: /opt/cray/pe/craype/2.6.2/bin/CC - works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- CMAKE_BUILD_TYPE is
-- Looking for sgemm_
-- Looking for sgemm_ - found
-- Found BLAS: BLAS_LIBRARIES-PLACEHOLDER-FOR-EMPTY-LIBRARIES
-- Looking for C++ include pthread.h
-- Looking for C++ include pthread.h - found
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Success
-- Found Threads: TRUE
-- LINALG libraries found:
-- OpenBLAS not found. We'll build it from source.
-- Found OpenMP_CXX: -fopenmp (found version "4.5")
-- Found OpenMP: TRUE (found version "4.5")
-- Configuring done
-- Generating done
-- Build files have been written to: /global/homes/f/friesen/NESAP/ASGarD/asgard.git/build_cce9
friesen@cori09:build_cce9>
```

**Comment by bcfriesen** [comment]
So this part:
```console
-- Looking for sgemm_
-- Looking for sgemm_ - found
-- Found BLAS: BLAS_LIBRARIES-PLACEHOLDER-FOR-EMPTY-LIBRARIES
```

is the new CMake behavior for handling LibSci, which it seems to do correctly in this case. But it looks like it's trying to build OpenBLAS anyway?
```console
-- OpenBLAS not found. We'll build it from source.
```

**Comment by bcfriesen** [comment]
Submitted #218 to fix this.

**Comment by grahamlopez** [comment]
closed by #218
